File: warrior/public/js/material_requests.js
frappe.ui.form.on("Material Request", {
refresh(frm) {

        if (!frm.doc.custom_supplier && !frm.doc.name) {
            frm.set_df_property("custom_supplier", "mandatory", 1);

            frm.set_df_property("items", "cannot_add_rows", true);
            frm.set_df_property("items", "hidden", 1);
        } else {
            frm.set_df_property("custom_supplier", "mandatory", 0);
            frm.set_df_property("items", "cannot_add_rows", false);
            frm.set_df_property("items", "hidden", 0);
        }
    },
  custom_supplier(frm) {
    if (!frm.doc.custom_supplier) return;
 frm.set_df_property("items", "cannot_add_rows", false);
            frm.set_df_property("items", "hidden", 0);
    // 🔹 Fetch supplier price list
    frappe.db.get_value(
      "Supplier",
      frm.doc.custom_supplier,
      "default_price_list"
    ).then(r => {
      if (r.message) {
        frm.set_value("buying_price_list", r.message.default_price_list || "");
      }
    });

    // 🔥 CLEAR OLD ITEMS (IMPORTANT)
    frm.clear_table("items");
    frm.refresh_field("items");

    // 🔹 Fetch allowed items
    frappe.call({
      method: "warrior.public.material_requests.items_by_supplier",
      args: { supplier: frm.doc.custom_supplier },
      callback(r) {

        const allowed = r.message || [];

        // ✅ Rebind query (fresh)
        frm.set_query("item_code", "items", function () {
          return {
            filters: [["Item", "name", "in", allowed]]
          };
        });

        // 🔄 Refresh grid UI
        frm.refresh_field("items");
      }
    });
  }
});

frappe.ui.form.on("Material Request Item", {
  item_code(frm, cdt, cdn) {
console.log(frm.doc.custom_supplier);
    // check parent field
    if (!frm.doc.custom_supplier) {

      frappe.msgprint("Please select the Supplier first");

      // clear the item_code user just selected
      frappe.model.set_value(cdt, cdn, "item_code", null);

      return;
    }
  }
});
frappe.ui.form.on("Material Request", {
    refresh(frm) {

        
        frm.page.remove_inner_button("Purchase Order", "Create");
        frm.page.remove_inner_button("Request for Quotation", "Create");
        frm.page.remove_inner_button("Supplier Quotation", "Create");

        if (!frm.doc.custom_supplier) return;

        frm.add_custom_button(
    __("Purchase Order"),
    () => {

        frappe.call({
            method: "warrior.public.material_requests.get_purchase_order_map",
            args: {
                source_name: frm.doc.name
            },
            callback: function(r) {

                let po = r.message;

                if (!po) return;

                // ✔ IMPORTANT: register doc in Frappe model
                let docs = frappe.model.sync(po);

                // ✔ get actual doc reference
                let doc = docs[0];

                // ✔ open form properly
                frappe.set_route("Form", doc.doctype, doc.name);
            }
        });

    },
    __("Create")
);
    }
});

File: warrior/public/material_requests.py
import frappe

@frappe.whitelist()
def items_by_supplier(supplier=None):
    """Get items filtered by supplier for Material Request Item field."""
    if not supplier:
        return []
    # items=frappe.get_all("supplier item table",filters={"supplier_id":supplier},pluck="parent")
    items=frappe.get_all("Item Supplier",filters={"supplier":supplier},pluck="parent")
    return items

@frappe.whitelist()
def get_purchase_order_map(source_name):

    mr = frappe.get_doc("Material Request", source_name)

    po = frappe.new_doc("Purchase Order")

    po.naming_series = "PUR-ORD-.YYYY.-"
    po.supplier = mr.custom_supplier
    po.company = mr.company
    po.transaction_date = frappe.utils.nowdate()
    po.schedule_date = mr.schedule_date
    po.set_warehouse = mr.set_warehouse
    po.buying_price_list = mr.buying_price_list
    po.tc_name = "Terms And Condition"

    po.custom_reference_type = "Material Request"
    po.custom_reference_name = mr.name

    for row in mr.items:
        po.append("items", {
            "item_code": row.item_code,
            "item_name": row.item_name,
            "qty": row.qty,
            "uom": row.uom,
            "rate": row.rate,
            "amount": row.amount,
            "warehouse": row.warehouse,
            "stock_uom":row.stock_uom,
            "conversion_factor":row.conversion_factor,
            "expense_account":row.expense_account
        })

    # ❌ DO NOT INSERT
    # po.insert()

    return po.as_dict()